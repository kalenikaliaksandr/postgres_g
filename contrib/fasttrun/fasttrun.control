comment = 'fast transaction-unsafe truncate'
default_version = '2.0'
module_pathname = '$libdir/fasttrun'
relocatable = true

