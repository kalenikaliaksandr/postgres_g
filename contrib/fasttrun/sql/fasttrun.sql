CREATE EXTENSION fasttrun;

create table persist ( a int );
insert into persist values (1);
select fasttruncate('persist');
insert into persist values (2);
select * from persist order by a;

create temp table temp1 (a int);
insert into temp1 values (1);

BEGIN;

create temp table temp2 (a int);
insert into temp2 values (1);

select * from temp1 order by a;
select * from temp2 order by a;

insert into temp1 (select * from generate_series(1,10000));
insert into temp2 (select * from generate_series(1,11000));

analyze temp2;
select relname,  relpages>0, reltuples>0 from pg_class where relname in ('temp1', 'temp2') order by relname;

select fasttruncate('temp1');
select fasttruncate('temp2');

insert into temp1 values (-2);
insert into temp2 values (-2);

select * from temp1 order by a;
select * from temp2 order by a;

COMMIT;

select * from temp1 order by a;
select * from temp2 order by a;

select relname,  relpages>0, reltuples>0 from pg_class where relname in ('temp1', 'temp2') order by relname;

select fasttruncate('temp1');
select fasttruncate('temp2');

select * from temp1 order by a;
select * from temp2 order by a;

select relname,  relpages>0, reltuples>0 from pg_class where relname in ('temp1', 'temp2') order by relname;
