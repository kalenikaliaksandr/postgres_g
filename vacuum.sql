delete from x where (c~>1)>0.1;
vacuum x;
