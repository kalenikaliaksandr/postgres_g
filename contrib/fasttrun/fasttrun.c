#include "postgres.h"

#include "access/genam.h"
#include "access/heapam.h"
#include "miscadmin.h"
#include "storage/lmgr.h"
#include "storage/bufmgr.h"
#include "catalog/namespace.h"
#include "utils/lsyscache.h"
#include "utils/builtins.h"
#include <fmgr.h>
#include <funcapi.h>
#include <access/heapam.h>
#include <catalog/pg_type.h>
#include <catalog/heap.h>
#include <commands/vacuum.h>
#include <utils/regproc.h>
#include <utils/varlena.h>

#ifdef PG_MODULE_MAGIC
PG_MODULE_MAGIC;
#endif

PG_FUNCTION_INFO_V1(fasttruncate);
Datum	fasttruncate(PG_FUNCTION_ARGS);
Datum
fasttruncate(PG_FUNCTION_ARGS) {
	text	*name=PG_GETARG_TEXT_P(0);
	char	*relname;
	List	*relname_list;
	RangeVar	*relvar;
	Oid		relOid;
	Relation	rel;
	bool	makeanalyze = false;

	relname = palloc( VARSIZE(name) + 1);
	memcpy(relname, VARDATA(name), VARSIZE(name)-VARHDRSZ);
	relname[ VARSIZE(name)-VARHDRSZ ] = '\0';

	relname_list = stringToQualifiedNameList(relname);
	relvar = makeRangeVarFromNameList(relname_list);
	relOid = RangeVarGetRelid(relvar, AccessExclusiveLock, false);

	if ( get_rel_relkind(relOid) != RELKIND_RELATION )
		elog(ERROR,"Relation isn't a ordinary table");

	rel = heap_open(relOid, NoLock);

	if ( !isTempNamespace(get_rel_namespace(relOid)) )
		elog(ERROR,"Relation isn't a temporary table");

	heap_truncate(list_make1_oid(relOid));

	if ( rel->rd_rel->relpages > 0 || rel->rd_rel->reltuples > 0 )
		makeanalyze = true;

	/*
	 * heap_truncate doesn't unlock the table,
	 * so we should unlock it.
	 */

	heap_close(rel, AccessExclusiveLock);

	if ( makeanalyze ) {
		VacuumParams	params;

		params.freeze_min_age = -1;
		params.freeze_table_age = -1;
		params.multixact_freeze_min_age = -1;
		params.multixact_freeze_table_age = -1;
		params.is_wraparound = false;
		params.log_min_duration = -1;

		vacuum(VACOPT_ANALYZE, NULL, relOid, &params, NULL,
				GetAccessStrategy(BAS_VACUUM), false);
	}

	PG_RETURN_VOID();
}
