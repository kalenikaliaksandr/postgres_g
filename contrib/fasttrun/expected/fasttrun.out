CREATE EXTENSION fasttrun;
create table persist ( a int );
insert into persist values (1);
select fasttruncate('persist');
ERROR:  Relation isn't a temporary table
insert into persist values (2);
select * from persist order by a;
 a 
---
 1
 2
(2 rows)

create temp table temp1 (a int);
insert into temp1 values (1);
BEGIN;
create temp table temp2 (a int);
insert into temp2 values (1);
select * from temp1 order by a;
 a 
---
 1
(1 row)

select * from temp2 order by a;
 a 
---
 1
(1 row)

insert into temp1 (select * from generate_series(1,10000));
insert into temp2 (select * from generate_series(1,11000));
analyze temp2;
select relname,  relpages>0, reltuples>0 from pg_class where relname in ('temp1', 'temp2') order by relname;
 relname | ?column? | ?column? 
---------+----------+----------
 temp1   | f        | f
 temp2   | t        | t
(2 rows)

select fasttruncate('temp1');
 fasttruncate 
--------------
 
(1 row)

select fasttruncate('temp2');
 fasttruncate 
--------------
 
(1 row)

insert into temp1 values (-2);
insert into temp2 values (-2);
select * from temp1 order by a;
 a  
----
 -2
(1 row)

select * from temp2 order by a;
 a  
----
 -2
(1 row)

COMMIT;
select * from temp1 order by a;
 a  
----
 -2
(1 row)

select * from temp2 order by a;
 a  
----
 -2
(1 row)

select relname,  relpages>0, reltuples>0 from pg_class where relname in ('temp1', 'temp2') order by relname;
 relname | ?column? | ?column? 
---------+----------+----------
 temp1   | f        | f
 temp2   | f        | f
(2 rows)

select fasttruncate('temp1');
 fasttruncate 
--------------
 
(1 row)

select fasttruncate('temp2');
 fasttruncate 
--------------
 
(1 row)

select * from temp1 order by a;
 a 
---
(0 rows)

select * from temp2 order by a;
 a 
---
(0 rows)

select relname,  relpages>0, reltuples>0 from pg_class where relname in ('temp1', 'temp2') order by relname;
 relname | ?column? | ?column? 
---------+----------+----------
 temp1   | f        | f
 temp2   | f        | f
(2 rows)

